File: itv/AMarotoChivite-ITV/AMarotoChivite-ITV-Java/src/main/java/FunctionalITVJava/base/ICrudITV.java
package FunctionalITVJava.base;

import FunctionalITVJava.models.Car;
import FunctionalITVJava.models.Vehicle;

import java.util.List;
import java.util.Map;

public interface ICrudITV<T> {
    List<T> getAll();

    List<Car> getOnlyCars();

    T getMoreModern();

    T getLessKm();

    List<T> searchByBrand(String brandToSearch);

    Double getAverageKmAllMoto();

    T getCarMoreAncientWithMoreTwoDoors();

    Map<String, Long> getNumByBrand();

    Map<String, Long> getNumAptosByVehicle();

    Map<String, Double> getByVehicleAverageYearFabricate();

    Map<String, List<Vehicle>> getAllGroupByBrand();

    List<T> getAllOrderByYearFabricate();

    List<T> getAllOrderByBrandDesc();

    Map<String, List<Vehicle>> getAllGroupByBrandAndSortedByKmDesc();
}

File: itv/AMarotoChivite-ITV/AMarotoChivite-ITV-Kotlin/src/main/kotlin/FunctionalITV/base/ICrudITV.kt
package FunctionalITV.base

import FunctionalITV.models.Vehicle

interface ICrudITV<T> {
    fun getAll(): List<T>

    fun getOnlyCars(): List<FunctionalITV.models.Car>

    fun getMoreModern(): T

    fun getLessKm(): T

    fun searchByBrand(brandToSearch: String): List<T>

    fun getAverageKmAllMoto(): Double

    fun getCarMoreAncientWithMoreTwoDoors(): T

    fun getNumByBrand(): Map<String, Int>

    fun getNumAptosByVehicle(): Map<String, Int>

    fun getByVehicleAverageYearFabricate(): Map<String, Int>

    fun getAllGroupByBrand(): Map<String, List<Vehicle>>

    fun getAllOrderByYearFabricate(): List<T>

    fun getAllOrderByBrandDesc(): List<T>

    fun getAllGroupByBrandAndSortedByKmDesc(): Map<String, List<Vehicle>>
}